{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dbf9fab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import loompy as lp\n",
    "from scipy import sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5f754a0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/pyscenic/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/anaconda3/envs/pyscenic/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4888 × 22430\n",
       "    obs: 'n_genes'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H5 = \"CTR9_snRNASeq/KO/filtered_feature_bc_matrix.h5\" \n",
    "\n",
    "adata = sc.read_10x_h5(H5)\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "# Basic QC filters (tweak as needed)\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "# Keep raw counts in a layer for later\n",
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "491777e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTR9_KO.loom'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loom_out = \"CTR9_KO.loom\"\n",
    "\n",
    "X = adata.layers[\"counts\"]\n",
    "if sparse.issparse(X):\n",
    "    X = X.T.tocsc()   # loom wants genes x cells\n",
    "else:\n",
    "    X = X.T\n",
    "\n",
    "row_attrs = {\"Gene\": np.array(adata.var_names)}\n",
    "col_attrs = {\"CellID\": np.array(adata.obs_names)}\n",
    "\n",
    "lp.create(loom_out, X, row_attrs, col_attrs)\n",
    "loom_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "70e96a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "LOOM_IN = \"CTR9_KO.loom\"\n",
    "\n",
    "# SCENIC resources\n",
    "TF_LIST = \"pyscenic_resources/allTFs_mm.txt\"\n",
    "DBS = [\n",
    "    \"pyscenic_resources/databases/mm10_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather\",\n",
    "    \"pyscenic_resources/databases/mm10_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.scores.feather\"\n",
    "]\n",
    "MOTIF_ANNOT = \"pyscenic_resources/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e09dffc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## in terminal:\n",
    "\n",
    "conda activate pyscenic\n",
    "\n",
    "arboreto_with_multiprocessing.py \\\n",
    "  CTR9_KO.loom \\\n",
    "  pyscenic_resources/allTFs_mm.txt \\\n",
    "  --method grnboost2 \\\n",
    "  --output CTR9_KO.adjacencies.tsv \\\n",
    "  --num_workers 8 \\\n",
    "  --seed 777"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyscenic",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
